### PR TITLE
Feat: add campaign pool

### LINKED ISSUE #58: Implement create_campaign Function with Unit Tests
The create_campaign_pool function is responsible for creating a new campaign contract, minting an NFT for the campaign owner, and associating the campaign with a unique campaign_pool_id. The function reads the token_giver_nft_contract_address, interacts with the ITokenGiverNftDispatcher to mint an NFT, retrieves the token_id, and uses IRegistryLibraryDispatcher to deploy the campaign contract. It then stores the campaign details and emits a CreateCampaign event.

```
      fn create_campaign_pool(
            ref self: ContractState,
            registry_hash: felt252,
            implementation_hash: felt252,
            salt: felt252,
            recipient: ContractAddress,
            campaign_pool_id: u256,
        ) -> ContractAddress {}
```

## Tasks:
* Implement the create_campaign_pool function to:
    * Read and increment the campaign pool count.
    * Retrieve the token_giver_nft_contract_address.
    * Mint an NFT for the campaign owner.
    * Get the minted token_id.
    * Deploy the campaign contract using IRegistryLibraryDispatcher::create_account.
    * Store campaign details in state.
    * Emit the CreateCampaign event.
    
## Acceptance Criteria:
* Ensure the create_campaign_pool function correctly initializes and stores campaign data.
* The function must successfully mint an NFT and associate it with the campaign.
* Campaign pool contract must be deployed correctly with the provided parameters.
* State updates (campaigns, campaign_nft_token, donation_count, etc.) should be validated.
* Ensure the correct event (CreateCampaignPool) is emitted with the expected details.

## Unit Tests:
* Test case: Successfully create a campaign and verify:
    * The campaign address is stored correctly.
    * The campaign owner matches the caller.
    * The NFT token_uri is correct.
    * The campaign details (e.g., campaign_id, token_id) match expected values.
* Test case: Ensure campaign count increments correctly.
* Test case: Validate the NFT minting process and token_id retrieval.
* Test case: Ensure contract deployment via create_account is successful.
* Test case: Verify the correct event (CreateCampaign) is emitted with expected parameters.

## Notes:
* The function currently assumes successful NFT minting and contract deployment; error handling should be considered.
* The get_caller_address() function is used for retrieving the campaign owner; ensure correctness in the context of execution.
* Consider adding additional edge case tests, such as handling duplicate campaign_id values or invalid NFT minting scenarios.

NOTE: pull from pool_campaign branch and submit your PR to pool_campaign

DURATION: 2 Days

### PR DESCRIPTION
**Pull request by Akshola00**
close #58

### CODE REVIEW

**Comment by mubarak23** [comment]
@Akshola00 GM , Please can you also add unit test for the function

File: src/campaign_pool.cairo [hunk 1]
```diff
@@ -132,7 +147,46 @@ mod CampaignPools {
...
+
+            self.campaign_pool.write(caller, campaign_details);
+            self
+                .emit(
+                    CreateCampaignPool {
+                        owner: caller,
+                        campaign_pool_address: campaign_address,
+                        token_id,
+                        campaign_pool_id: campaign_pool_count.try_into().unwrap(),
+                        nft_token_uri: " ",
```
Comment (mubarak23) [inline src/campaign_pool.cairo:182]: you can get token url by calling `get_token_uri` on the NFT Contract function

File: src/campaign_pool.cairo [hunk 2]
```diff
@@ -144,10 +198,9 @@ mod CampaignPools {
             campaign_pool_address: ContractAddress,
             amount: u256
         ) {}
-
-        fn upgrade(ref self: ContractState, new_class_hash: ClassHash) {
```
Comment (mubarak23) [inline src/campaign_pool.cairo:148]: we should not comment this out, we want to be able to upgrade the contract

File: src/interfaces/ICampaignPool.cairo [hunk 3]
```diff
@@ -23,5 +23,5 @@ pub trait ICampaignPool<TState> {
         campaign_pool_address: ContractAddress,
         amount: u256
     );
-    fn upgrade(ref self: TState, new_class_hash: ClassHash);
+    // fn upgrade(ref self: TState, new_class_hash: ClassHash);
```
Comment (mubarak23) [inline src/interfaces/ICampaignPool.cairo:26]: same with the interface definition, we want to be able to upgrade the contract

File: src/campaign_pool.cairo (hunk 1 above)
Comment (mubarak23) [reply src/campaign_pool.cairo:182]: i.e `let token_uri = token_giver_dispatcher.get_token_uri(token_id);`

**Review by mubarak23 — CHANGES_REQUESTED** [review]

**Comment by Akshola00** [comment]
``` task
Store campaign details in state.
```
```
campaign_pool: Map<ContractAddress, CampaignPool>,

 campaign_pool_applications: Map<
            ContractAddress, (ContractAddress, u256)
        >, // map<Campaign Address, (campaign pool address, amount)>
        campaign_pool_count: u16,
        campaign_pool_nft_token: Map<
            ContractAddress, (ContractAddress, u256)
        >, // (recipient, (campaign_address, token_id));

```
@mubarak23 do i store the campaign details in any other state apart from campaign pool
currently im storing in only campaign pool

**Review by mubarak23 — CHANGES_REQUESTED** [review]
aside this one change, this PR is solid

**Comment by Akshola00** [comment]
will do so now

**Comment by Akshola00** [comment]
@mubarak23 ptal

**Comment by mubarak23** [comment]
@Akshola00 check the test function, one of them is failing due to struct field name change, look the github action result for test for more insight

**Comment by Akshola00** [comment]
@mubarak23 Thanks, it has been changed, i didn't update after i changed the struct field but now i have done so
struct and tests have been updated now and ci is passing

**Review by mubarak23 — APPROVED** [review]
lgfm
